### PR TITLE
[JENKINS-23152] More robust handling of build number collisions

### PR DESCRIPTION
**Pull request by jglick**
See [JENKINS-23152](https://issues.jenkins.io/browse/JENKINS-23152) as defended against originally in #1379. If `nextBuildNumber` is stale or the in-memory build number list fails to match what is on disk in some other way, you can get errors like

```
SEVERE	hudson.model.Executor#run: Executor #-1 for Built-In Node: Unexpected executor death
java.lang.IllegalStateException: JENKINS-23152: …/jobs/prj/builds/123 already existed; will not overwrite with prj #123
	at hudson.model.RunMap.put(RunMap.java:193)
	at jenkins.model.lazy.LazyBuildMixIn.newBuild(LazyBuildMixIn.java:184)
Caused: java.lang.LinkageError: JENKINS-23152: …/jobs/prj/builds/123 already existed; will not overwrite with prj #123
	at jenkins.model.lazy.LazyBuildMixIn.newBuild(LazyBuildMixIn.java:192)
	at jenkins.model.ParameterizedJobMixIn$ParameterizedJob.createExecutable(ParameterizedJobMixIn.java:505)
	at jenkins.model.ParameterizedJobMixIn$ParameterizedJob.createExecutable(ParameterizedJobMixIn.java:323)
	at hudson.model.Executor$1.call(Executor.java:370)
	at hudson.model.Executor$1.call(Executor.java:352)
	at hudson.model.Queue._withLock(Queue.java:1456)
	at hudson.model.Queue.withLock(Queue.java:1312)
	at hudson.model.Executor.run(Executor.java:352)
```

which prevent a new build from being run at all. This patch at least retries with the newly updated `nextBuildNumber` until it finds an open slot.

#2439 also added some robustness during `onLoad`, which cannot be used reliably here since that presumes `maxNumberOnDisk` is freshly computed from actually inspecting the `builds` directory.

### Testing done

Manually decrementing `nextBuildNumber` on disk. Not sure this is worth an automated test.

### Proposed changelog entries

- Robustness improvement regarding build number collisions.

### Maintainer checklist

Before the changes are marked as `ready-for-merge`:

- [x] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [x] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [x] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [x] Proper changelog labels are set so that the changelog can be generated automatically.
- [x] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [x] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/7523"></a>

### CODE REVIEW

File: core/src/main/java/jenkins/model/lazy/LazyBuildMixIn.java
Comment (jglick) [inline core/src/main/java/jenkins/model/lazy/LazyBuildMixIn.java:188]: Yes this is (almost) tail-recursion and could be rewritten to use a loop. OTOH if there is actually a list of thousands of builds on disk newer than the purported `nextBuildNumber`, a `StackOverflowError` is arguably no worse than the previous behavior, and arguably preferable to going into an endless loop. Could tweak to be a bounded loop etc.

File: core/src/main/java/jenkins/model/lazy/LazyBuildMixIn.java
Comment (jglick) [inline core/src/main/java/jenkins/model/lazy/LazyBuildMixIn.java:198]: Only tangentially related, but note the inappropriate use of `LinkageError` in the PR description. This seems to be because #2125 added a multicatch which #5483 translated incorrectly.

**Review by timja — APPROVED** [review]

**Review by NotMyFault — APPROVED** [review]

**Comment by NotMyFault** [comment]
/label ready-for-merge

---

This PR is now ready for merge. We will merge it after ~24 hours if there is no negative feedback.
Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process.
Thanks!

**Pull request merged**
